packages/icalendar/type/Recur.pkl

```diff
@@ -0,0 +1,198 @@
+//===----------------------------------------------------------------------===//
+// Copyright © 2025 Apple Inc. and the Pkl project authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//===----------------------------------------------------------------------===//
+/// A recurrence rule.
+///
+/// https://datatracker.ietf.org/doc/html/rfc5545#section-3.3.10
+module icalendar.Recur
+
+import "Date.pkl"
+import "DateTime.pkl"
+
+typealias Frequency = "SECONDLY"|"MINUTELY"|"HOURLY"|"DAILY"|"WEEKLY"|"MONTHLY"|"YEARLY"
+
+typealias WeekdayString = "SU"|"MO"|"TU"|"WE"|"TH"|"FR"|"SA"
+
+class Weekday {
+  weekday: WeekdayString
+  weeknum: Int(this != 0)?
+
+  function toString(): String = List(
+    weeknum?.ifNonNull((n) -> if (n as Int > 0) "+" else ""),
+    weeknum?.toString(),
+    weekday
+  ).filterNonNull().join("")
+}
+
+typealias ByWeekday = WeekdayString|* Weekday
+
+dtstart: (
+  * DateTime(utc.implies(tzid == null))
+  |Date(tzid == null)
+  )?
+
+tzid: String?
+
+/// The FREQ rule part identifies the type of recurrence rule.
+freq: Frequency(
+    !(count != null && until != null) || throw("count and until may not both be set")
```

Comment:
[nit] move this constraint to either `count` or `until`, because this doesn't have anything to do with `freq`